Refactor `Doctype`

gsteel opened the pull request:
Instead of introducing a V3 helper here _(Why did I do that?)_, I've re-applied the changes to the existing Doctype helper.

- Removes inheritance
- Adds types
- Adds the `final` keyword
- Removes all runtime mutation
- Changes default doctype to HTML5
- Removes the ability to use custom doctypes
- Removed methods
  - `setView`
  - `getView`
  - `setDoctype`
  - `getDoctype`
  - `getDoctypes`
  - `unsetDoctypeRegistry`
- implements `Stringable`

A number of tests rely on runtime mutation of the current doctype - these have been patched in a pretty ugly way for the time-being.

All of these helpers will eventually be refactored to use constructor injection of the Doctype helper so they can modify their output based on the configured doctype.

Examples are `HeadScript`, `HeadStyle` etc.

gsteel committed:
Initial RFC for Doctype Replacement Helper

Signed-off-by: George Steel <george@net-glue.co.uk>

Ocramius commented on src/Helper/DoctypeV3.php:
```diff
@@ -0,0 +1,112 @@
+        return $this->doctypeDeclaration();
+    }
+
+    /** @return non-empty-string */
```
IMO `key-of<self::$docTypeDefinitions>`

gsteel replied:
Nice - Didn't know that!

Ocramius commented on src/Helper/DoctypeV3.php:
```diff
@@ -0,0 +1,112 @@
+    public const HTML5               = 'HTML5';
+
+    /** @var array<string, non-empty-string> */
+    private static $docTypeDefinitions = [
```
Can be a constant

Ocramius commented on src/Helper/DoctypeV3.php:
```diff
@@ -0,0 +1,112 @@
+    /**
+     * Is doctype XHTML?
+     */
+    public function isXhtml(?string $doctypeId = null): bool
```
Also here, let's restrict these input types as much as possible

gsteel committed:
Refine types for arguments and return types and change static property to a constant

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Refactor the `Doctype` helper

- Removes inheritance
- Adds types
- Adds the `final` keyword
- Removes all runtime mutation
- Changes default doctype to HTML5
- Removes the ability to use custom doctypes
- Removed methods
  - `setView`
  - `getView`
  - `setDoctype`
  - `getDoctype`
  - `getDoctypes`
  - `unsetDoctypeRegistry`
- implements `Stringable`

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Fix tests that rely on runtime doctype mutation

Signed-off-by: George Steel <george@net-glue.co.uk>

Ocramius commented on src/Helper/Doctype.php:
```diff
@@ -32,199 +31,95 @@ class Doctype extends AbstractHelper
-     * Register the default doctypes we understand
-     *
-     * @return void
+     * @param DoctypeID|string|null $doctypeId
```
Thr `|string` kinda kills all type safety introduced here though

gsteel replied:
Yes - this was mainly so that the defensive check/exception was worth doing. I figured the check was worth retaining because of less precise inputs from a view script context.

Theoretically, no-one will interact with that method anyhow, unless they are using the helper as a lookup table, and it'll crash for invalid input.

gsteel replied:
Am hoping that I can ship a Psalm plugin and collector (or whatever they're called) so we can get SA on .phtml files BTW

gsteel committed:
Check for doctype configuration in traditional MVC locations

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Update doctype documentation for v3

Signed-off-by: George Steel <george@net-glue.co.uk>

froschdesign commented on docs/book/v3/helpers/doctype.md:
```diff
@@ -18,38 +17,31 @@ The `Doctype` helper allows you to specify one of the following types:
 
-$doctypeHelper = new Doctype();
-$doctypeHelper->doctype('XHTML1_STRICT');
+$helper = new Doctype(Doctype::HTML5);
```
HTML5 is the default doctype therefore it should be omitted.
A second example with a different doctype would be good.

froschdesign commented on docs/book/v3/helpers/doctype.md:
````diff
@@ -79,13 +71,14 @@ return [
 ];
 ```
 
+!NOTE The default doctype is HTML 5 when no configuration is specified.
````
The syntax is wrong but I would explain this with a second example, see above.

> ## Basic Usage
> 
> Start the admonition with one of the types in capital letters followed by a
> colon.
> The type corresponds with a specific style.
> The type should be followed by the content of the admonition.
> 
> ```markdown
> NOTE: An example of a note.
> ```
>
> ## Type of Admonitions
> 
> | Title / Topic             | Type    |
> |---------------------------|---------|
> | New Feature               | TIP     |
> | Deprecation Warning       | CAUTION |
> | Installation Requirements | MISSING |
> | General Notes             | NOTE    |
> | Security Warning          | WARNING |
